File: leetcode/124-Binary-Tree-Maximum-Path-Sum/tina.swift
class Solution {
    func maxPathSum(_ root: TreeNode?, max: inout Int) -> Int {
        if root == nil {
            return 0
        }
        
        let node: TreeNode! = root
        let l = maxPathSum(node.left, max: &max)
        let r = maxPathSum(node.right, max: &max)
        
        let sum = Swift.max(node.val, l + node.val, r + node.val)
        max = Swift.max(max, sum, l + node.val + r)
        
        return sum
    }
    
    func maxPathSum(_ root: TreeNode?) -> Int {
        var max = Int(Int32.min)
        maxPathSum(root, max: &max)
        
        return max
    }
}

File: leetcode/235-Lowest-Common-Ancestor-of-a-Binary-Search-Tree/tina.cpp
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == NULL) return NULL;
        
        int dp = p->val - root->val;
        int dq = q->val - root->val;
        
        if (dp > 0 && dq > 0) {
            return lowestCommonAncestor(root->right, p, q);
        } else if (dp < 0 && dq < 0) {
            return lowestCommonAncestor(root->left, p, q);
        }
        
        return root;
    }
};
